pymongo/asynchronous/mongo_client.py

```diff
@@ -1117,14 +1218,22 @@ def options(self) -> ClientOptions:
 
     def __eq__(self, other: Any) -> bool:
         if isinstance(other, self.__class__):
-            return self._topology == other._topology
+            if hasattr(self, "_topology") and hasattr(other, "_topology"):
+                return self._topology == other._topology
+            else:
+                raise InvalidOperation(
+                    "Cannot compare client equality until both clients are connected"
+                )
         return NotImplemented
 
     def __ne__(self, other: Any) -> bool:
         return not self == other
 
     def __hash__(self) -> int:
-        return hash(self._topology)
+        if hasattr(self, "_topology"):
+            return hash(self._topology)
+        else:
+            raise InvalidOperation("Cannot hash client until it is connected")
```

Comment:
It seems odd to disallow comparison and hashing without "connecting" first. I wonder if there's a better way to define these without relying on SRV or TXT resolution. Like could be just compare some subset of MongoClient arguments?